### PR TITLE
Make IndentationCheck accept right parantheses of annotation alone in…

### LINKED ISSUE #1349: Indentation on annotation parameters is wrong
When using `<property name="tabWidth" value="4"/><module name="Indentation"/>`, during build checkstyle 6.8.1 complains about:
```
    @Cascade(
        org.hibernate.annotations.CascadeType.DELETE_ORPHAN
    )
```
with:
```
ClickConfig.java:77: ')' have incorrect indentation level 4, expected level should be 8.
```

But it should accept such formatting as the right parenthes closes a block that started on indentation level 4, so it should also be on 4, not on 8.

### PR DESCRIPTION
**Pull request by liscju**
… line indent same as annotation begin, fixes #1349

### CODE REVIEW

**Comment by mkordas** [comment]
@liscju - thanks for the long-awaited fix. Can you also check whether this commit resolves #553?

Moreover, I believe we need couple more unit tests of complex annotations and diff before/after changes on Spring/Hibertate projects as they use annotations heavily.

**Comment by liscju** [comment]
@mkordas it is not resolving #553 because this fix only assumes that closing bracket ')' of annotation can be on the same line indent as annotation opening bracket ')' indent- nothing else. It was suggested in #553 that we should treat opening/closing bracket like scope - and it could be good idea apart from the fact that it is more complex situation - for example it is perfectly good annotation:

```java
@DifficultAnnotation({
        @MyType(value = Boolean.class, name = "boolean"), 
        @MyType(value = String.class, name = "string") }) 
```

and this:

```java
@DifficultAnnotation({
        @MyType(value = Boolean.class, name = "boolean"), 
        @MyType(value = String.class, name = "string") 
}) 
```

to get even more complex:

```java
@DifficultAnnotation({
        @MyComplexType({
                @MyInnerType(value = String.class, name = "string") 
         }), 
        @MyType(value = String.class, name = "string") 
}) 
```
 seems legit , like: 

```java
@DifficultAnnotation({
        @MyComplexType({
                @MyInnerType(value = String.class, name = "string") }), 
        @MyType(value = String.class, name = "string") 
}) 
```

Maybe checkAnnotationIndentation in LineWrappingHandler (https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LineWrappingHandler.java#L206) should be rewritten to use new class AnnotationBlockHandler extending BlockParentHandler(https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/BlockParentHandler.java) where such complex indentation could be handled. I could try to implement this, but this would be much more complex change than little fix and i would like to know before if there is any chance for such a solution to be approved

**Comment by romani** [comment]
@mkordas , are you ok with that changes ? are you finish code-review ?

**Comment by romani** [comment]
@mkordas , can you finish code-review for this PR?

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheckTest.java [hunk 1]
```diff
@@ -293,6 +293,25 @@ public void testDifficultAnnotations() throws Exception {
...
+        checkConfig.addAttribute("arrayInitIndent", "4");
+        checkConfig.addAttribute("basicOffset", "4");
+        checkConfig.addAttribute("braceAdjustment", "0");
+        checkConfig.addAttribute("caseIndent", "4");
+        checkConfig.addAttribute("forceStrictCondition", "false");
+        checkConfig.addAttribute("lineWrappingIndentation", "4");
+        checkConfig.addAttribute("tabWidth", "4");
+        checkConfig.addAttribute("throwsIndent", "4");
+
+        final String[] expected = {
```
Comment (mkordas) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheckTest.java]: we recently changed empty array initalizers to `ArrayUtils.EMPTY_STRING_ARRAY`

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheckTest.java [hunk 2]
```diff
@@ -293,6 +293,25 @@ public void testDifficultAnnotations() throws Exception {
...
+        };
+
+        verifyWarns(checkConfig, getPath("indentation/InputAnnotationClosingBracketEndInSameIndentationAsOpening.java"), expected);
```
Comment (mkordas) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheckTest.java]: let's keep our lines shorter

File: src/test/resources/com/puppycrawl/tools/checkstyle/indentation/InputAnnotationClosingBracketEndInSameIndentationAsOpening.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ *                                                                            //indent:1 exp:1
+ * arrayInitIndent = 4                                                        //indent:1 exp:1
+ * basicOffset = 4                                                            //indent:1 exp:1
+ * braceAdjustment = 0                                                        //indent:1 exp:1
+ * caseIndent = 4                                                             //indent:1 exp:1
+ * forceStrictCondition = false                                               //indent:1 exp:1
+ * lineWrappingIndentation = 4                                                //indent:1 exp:1
+ * tabWidth = 4                                                               //indent:1 exp:1
+ * throwsIndent = 4                                                           //indent:1 exp:1
+ *                                                                            //indent:1 exp:1
```
Comment (mkordas) [inline src/test/resources/com/puppycrawl/tools/checkstyle/indentation/InputAnnotationClosingBracketEndInSameIndentationAsOpening.java]: I know it's probably copy&paste, but these empty lines are not needed here

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheckTest.java [hunk 4]
```diff
@@ -293,6 +293,25 @@ public void testDifficultAnnotations() throws Exception {
...
+        final DefaultConfiguration checkConfig = createCheckConfig(IndentationCheck.class);
+
+        checkConfig.addAttribute("arrayInitIndent", "4");
+        checkConfig.addAttribute("basicOffset", "4");
+        checkConfig.addAttribute("braceAdjustment", "0");
+        checkConfig.addAttribute("caseIndent", "4");
+        checkConfig.addAttribute("forceStrictCondition", "false");
+        checkConfig.addAttribute("lineWrappingIndentation", "4");
+        checkConfig.addAttribute("tabWidth", "4");
+        checkConfig.addAttribute("throwsIndent", "4");
```
Comment (mkordas) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheckTest.java]: I don't thinks some of this settings matter for our case. For sure we're not testing anything related to `throws`, so we can rely on defaults if possible.

File: src/test/resources/com/puppycrawl/tools/checkstyle/indentation/InputAnnotationClosingBracketEndInSameIndentationAsOpening.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+@interface SimpleType { //indent:0 exp:0
+    Class<?> value(); //indent:4 exp:4
+} //indent:0 exp:0
+
+@SimpleType( //indent:0 exp:0
+    value = Boolean.class //indent:4 exp:4
+) //indent:0 exp:0
+public class InputAnnotationClosingBracketEndInSameIndentationAsOpening { //indent:0 exp:0
+} //indent:0 exp:0
```
Comment (mkordas) [inline src/test/resources/com/puppycrawl/tools/checkstyle/indentation/InputAnnotationClosingBracketEndInSameIndentationAsOpening.java]: This input is about `AnnotationClosingBracketEndInSameIndentationAsOpening`. For sure we can think of some more annotations for this file proving that functionality works correctly, e.g. something like that:
```java
@SimpleType( value = Boolean.class
) 
```

**Comment by liscju** [comment]
@mkordas 

1) DONE - Array Empty Initializer changed as requested
2) DONE - Line divided
3) DONE - DefaultConfiguration removed options:
checkConfig.addAttribute("arrayInitIndent", "4");
checkConfig.addAttribute("braceAdjustment", "0");
checkConfig.addAttribute("caseIndent", "4");
checkConfig.addAttribute("lineWrappingIndentation", "4");
checkConfig.addAttribute("throwsIndent", "4");

4) DONE - Removed useless header from InputAnnotationClosingBracketEndInSameIndentationAsOpening

5) DONE - Case added to InputAnnotationClosingBracketEndInSameIndentationAsOpening:
@SimpleType( value = Boolean.class
)

**Comment by mkordas** [comment]
@liscju, couple of final remarks:
1. Use our new commit message format with issue in name and try to keep in up to 70 chars long with additional description below
2. Fix typo in commit message (`parantheses`)
3. Please be consistent and use everywhere word 'parentheses' instead of 'bracket'
4. Let's add some negative test with expected violation if annotation closing parentheses has wrong indentation

**Comment by liscju** [comment]
@mkordas 
1) DONE - commit message renamed to "Issue #1349: Fix IndentationCheck to accept proper right parenthesis of annotation"

2) DONE

3) DONE, changed:

* test method to testAnnotationClosingParenthesisEndInSameIndentationAsOpening
 * input file to InputAnnotationClosingParenthesisEndInSameIndentationAsOpening

3) DONE

4) DONE, above class InputAnnotationClosingParenthesHasBadIndentation in InputAnnotationClosingParenthesisEndInSameIndentationAsOpening

**Comment by mkordas** [comment]
Thanks.

1. Just after merge of #2132 please rebase over latest sources to have TeamCity build passing
2. Minor: should be InputAnnotationClosingParenthesisEnd**s**InSameIndentationAsOpening
3. Maybe I'm too cautious here (and sorry for not mentioning it earlier), but it would be nice to have test where annotion is on something different than class and has initial indention level greater than 0 - what do you think?

After the above I'm OK to merge.

**Comment by romani** [comment]
to make teamcity pass , just rebase on latest master

**Comment by liscju** [comment]
@mkordas 
1) DONE
2) DONE
3) DONE, added example:

```java
class InputAnnotationClosingParenthesHasBadIndentation { //indent:0 exp:0
    @SimpleType( value = Boolean.class //indent:4 exp:4
        ) //indent:8 exp:4 warn
    Boolean booleanField; //indent:4 exp:4
} //indent:0 exp:0
```

Let me know if i should add another test cases.

P.S. More probably i am too "cautious" about adding new test cases - i'm always worring that too much of them makes browsing tests not clear and it will distract future reader.

**Comment by mkordas** [comment]
All right, code looks good, so the final thing we need to do is to compare reports before and after fix. I would choose any library with annotations used (e.g. Spring Framework) and analyze all the differences whether they are expected or not. Can you prepare them? Ask for assistance if needed.

I'd like also to validate how much false-positives will be fixed after your changes will be merged.

**Comment by romani** [comment]
> I would choose any library with annotations used (e.g. Spring Framework) and analyze all the differences whether they are expected or not

please run against  all repositories that we have in checksyle-tester and compare result before and after, I doubt there would be too much difference in violations.

**Comment by liscju** [comment]
In Spring Framework count of warning before/after is about 1084723/1084720, is there a simple way to find the difference between checkstyle reports instead of searching it manually?

**Comment by mkordas** [comment]
@liscju, I was using just a unix `diff` for that

**Comment by romani** [comment]
@liscju , please play a bit with Indentation config to catch what settings suits more to Spring sources to minimize amount of warnings. You can place result files to the web and compare them by online diff services.

**Comment by romani** [comment]
release 6.11will be on this Sunday (27th of September), it would be good to finish this PR before release.

Please let us know if you have problems with testing - we could help.

**Comment by liscju** [comment]
There were only 3 differences between before/after:

1) https://github.com/spring-projects/spring-framework/blob/v4.1.6.RELEASE/spring-context/src/test/java/org/springframework/context/annotation/ComponentScanAnnotationTests.java#L60

Before fix checkstyle raises "have incorrect indentation level 0, expected level should be 4"

2) https://github.com/spring-projects/spring-framework/blob/v4.1.6.RELEASE/spring-context/src/test/java/org/springframework/context/annotation/ComponentScanAnnotationIntegrationTests.java#L345

Before fix checkstyle raises "have incorrect indentation level 1, expected level should be 4"

3) https://github.com/spring-projects/spring-framework/blob/v4.1.6.RELEASE/spring-context/src/test/java/org/springframework/context/annotation/ComponentScanAnnotationIntegrationTests.java#L352

Before fix checkstyle raises "have incorrect indentation level 1, expected level should be 4"

Now i have hesitation if it is working correctly, it seems pretty impossible that in such a huge project like spring-framework ( which uses annotation heavily)
there were so few occurences of this bug. I am interested what do you think about it.

**Comment by mkordas** [comment]
@liscju, the difference looks good - all 3 violations were false positives. At least we are sure that we fix something and do not introduce any new problems. This is additionally confirmed by unit tests. If someone will spot more false-positives, new issue can be opened. @romani, I think we are good to merge.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LineWrappingHandler.java [hunk 6]
```diff
@@ -221,11 +221,13 @@ private void checkAnnotationIndentation(DetailAST atNode,
         while (firstNodesOnLines.size() > 1) {
             final DetailAST node = itr.next();
 
-            if (node.getLineNo() < lastAnnotationLine
-                    || node.getLineNo() == lastAnnotationLine) {
+            if (node.getLineNo() <= lastAnnotationLine) {
                 final DetailAST parentNode = node.getParent();
+                final boolean isCurrentNodeCloseAnnotationAloneInLine =
+                        node.getLineNo() == lastAnnotationLine && node == lastAnnotationNode;
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LineWrappingHandler.java]: Codacy Issue found: [Use equals() to compare object references.](https://www.codacy.com/app/checkstyle/checkstyle/file/1008404572/issues/source?bid=2193633&fileBranchId=2739117#l227)

Comment (liscju) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LineWrappingHandler.java]: Fixed, use equals() now

**Comment by romani** [comment]
@liscju ,

It is always better to provide two reports to let us validate that you are right or not. Right now I am not sure what Check is doing now on these cases. You shared only previous violations.
If for first case I can presume that we give no violations but for other I need to be sure what message we are printing ( "ComponentScanAnnotationIntegrationTests.java#L345" and "ComponentScanAnnotationIntegrationTests.java#L352")

**Comment by liscju** [comment]
@romani I totally agree with u, i wanted to do this with reports generated by spring-framework but they were huge. So huge that i encountered 100MB limitation for sharing single file in github.com. Probably as you mentioned by tweaking IndentationCheck parameters i could filter useless cases, but i was afraid that i would filtered out changes before/after fix too. I hesitated to share diff of those files but this wouldn't be any more convincing than just comments here. I would like to hear from you what to do in situation like this, especially because few times already i has this problem with report on spring-framework. 

For 2) and 3) after fix it doesnt show any violation in those lines and honestly im not sure if this is right.

**Comment by romani** [comment]
> For 2) and 3) after fix it doesnt show any violation in those lines and honestly im not sure if this is right.

yes, it is not right, as according to issue, "(" need to be under "@" (the same indentation) as you have in your case "1)"


> So huge that i encountered 100MB limitation for sharing single file in github.com.

you could remove most part of sources from validation of Springframework validation, see example (last option in config line) at https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/projects-to-test-on.properties#L17. To have multiple excludes you can separate them by "," (that will pass to checkstyle-maven-plugin as exclude option value https://maven.apache.org/plugins/maven-checkstyle-plugin/check-mojo.html#excludes)

Example of few folders exclude: "#openjdk|hg|http://hg.openjdk.java.net/jdk7/jdk7/jdk/||**/sun/**/*,**/javax/**/*,**/java/**/*"

This will minimize amount of files to validate and as result HTML report size

**Comment by romani** [comment]
@liscju , can we finish this PR ? do you need help.

It would be very helpful to finish this PR as we plan to resolve all other issue of this Check (changes will be conflicting :( )

**Comment by romani** [comment]
@liscju ,

release 6.12 is planned to be at 31 of October.
it would very good to finish this PR before release. Do you have time to finish PR ?

**Comment by romani** [comment]
@liscju , do you have time to finish this PR ?

If you do not have time for now to continue, it is completely ok to let us know and we will finish it, or another contributor could finish it.
I just do not want this issue be blocked by incomplete implementation.

**Comment by romani** [comment]
> So huge that i encountered 100MB limitation for sharing single file in github.com.

There is a project - https://github.com/attatrol/ahsm that is intended to address this issue and help do testing easily. Please use it. We are so close to finish this PR.

**Comment by romani** [comment]
@mkordas , looks like we lost contact with contributor. Can you finish this PR ? looks like only testing is left.

**Comment by liscju** [comment]
@romani sory for lack of responding. I have no time right now for finishing this issue, so somebody else can take this.

**Comment by romani** [comment]
@liscju , thanks you for you time to spend on this issue, in future do not hesitate to let us know that do not have time any more. It is ok. We just do not want issues/fixes to be hanging uncompleted for long time.

see you later on in checkstyle :).

**Comment by romani** [comment]
checkstyle/contribution#40 is fixed, size of site report is minimized now.

**Comment by mkordas** [comment]
@romani I did the testing and quality is good here. No new false-positives. Here's the diff between reports over Guava, Spring and Hibernate: http://mkordas.github.io/indentation-right-paren/checkstyle_merged.html

Can we merge this PR and maybe release 6.14.2?

**Comment by romani** [comment]
@liscju , mkordas generated a report with `-Xmx5G` option.

> Can we merge this PR and maybe release 6.14.2?

No need, release takes too much time, it is not urgent fix, it is already hanging for 6 month, +1 month delay is ok.

**Comment by mkordas** [comment]
@romani 
> it is not urgent fix

Right now for me this is blocker to introduce Indentation check in one of my projects, that's why I sacrificed entire evening to do the testing :)

**Comment by romani** [comment]
I might miss smth  but issue is about to avoid violation on code like:

```
    @Cascade(
        org.hibernate.annotations.CascadeType.DELETE_ORPHAN
    )
```

and in your report we have 3 violations in "site-after".

Example:
hibernate-orm/hibernate-envers/src/test/java/org/hibernate/envers/test/entities/manytomany/biowned/ListBiowning2Entity.java
Anchor 	Severity 	Category 	Rule 	Message 	Line 	Site
tity.1 	 Warning 	indentation 	Indentation 	')' have incorrect indentation level 1, expected level should be 4. 	54 	site-after
tity.2 	 Warning 	indentation 	Indentation 	')' have incorrect indentation level 1, expected level should be 8. 	54 	site-before

So, we changed expected level from 8 to 4 in message. 
But author of issue do not expect any violation on such code. BTW "have incorrect indentation level 1" is incorrect as it should be "have incorrect indentation level 4" at least.

**Comment by mkordas** [comment]
@romani It's because these lines use tabs instead of spaces. Violations depend on how `tabWidth` is configured.

**Comment by romani** [comment]
please share your config.

**Comment by mkordas** [comment]
@romani config is `tabSize=4`, so indeed there should be no violation here. But I've traced this problem to be independent bug, please see #2795. I don't think we have blocker here.

**Comment by romani** [comment]
ok, I am ok to merge, but CI is not happy with a code, see PR #2799 .

@liscju , can you rebase your code over latest code ?

> Right now for me this is blocker to introduce Indentation check in one of my projects, that's why I sacrificed entire evening to do the testing :)

@mkordas , after merge,  there will be a strict SNAPSHOT version - https://oss.sonatype.org/content/repositories/snapshots/com/puppycrawl/tools/checkstyle/6.15-SNAPSHOT/ , is it enough for you ?

**Comment by liscju** [comment]
@romani rebased and fixed some style error

**Comment by mkordas** [comment]
@romani 
> @mkordas , after merge, there will be a strict SNAPSHOT version - https://oss.sonatype.org/content/repositories/snapshots/com/puppycrawl/tools/checkstyle/6.15-SNAPSHOT/ , is it enough for you ?

Not really, I can't use other repos than Maven central in my work, but thanks a lot for merging!

**Comment by romani** [comment]
@mkordas ,

current release notes - https://travis-ci.org/checkstyle/checkstyle/jobs/100113661#L213
Even we have "New", it is still ok to do hotfix, as we do not introduce anything new from configuration point of view, and it could be classified as "bug" too.

In respect to your contributions to the our project I could do hot fix release. It would occupy whole hours of day that I could spent for a project. Release process is still not ideal and far from simple binaries publication to maven, even I do optimizations each month to make it more automated. 

Please confirm that you need a hotfix and would rather not wait till end of month.

**Comment by mkordas** [comment]
@romani it is not super-urgent for me, just nice to have and I've seen bit more releases happening recently (e.g. 6.11.2), so I haven't realized there's still so much work here. Maybe we could make a deal and do that release together, share some tasks? Actually I may need to take over entire release once at some point in the future, e.g. you going on holidays :)

**Comment by tunix** [comment]
Hi, I've `maven-checkstyle-plugin:3.0.0` (which seems to be using checkstyle 6.18) added to my project. I came across this problem with below code:

```
    @Caching(put = {
        @CachePut(key = "#foo"),
        @CachePut(key = "#bar")
    })
```

Above code causes the error below:

```
... (indentation) Indentation: '}' have incorrect indentation level 4, expected level should be 8.
```

Modifying my checkstyle config like below solved the issue:

```
        <module name="Indentation">
            <property name="lineWrappingIndentation" value="0"/>
        </module>
```

I guess this shouldn't have been necessary if the changes in this PR is effective, right? What am I missing?

**Comment by romani** [comment]
@tunix , please review open issues in indentation check, if nothing similar to your case - open new issue

**Comment by Van0SS** [comment]
@tunix They actually did fix the issue. I changed version of checkstyle to latest (8.14 atm) and didn't have this problem anymore. Maven config will look like:
```
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-checkstyle-plugin</artifactId>
                <version>3.0.0</version>
                <executions>
                    <execution>
                         YOUR_EXECUTION_SETUP
                    </execution>
                </executions>
                <dependencies>
                    <dependency>
                        <groupId>com.puppycrawl.tools</groupId>
                        <artifactId>checkstyle</artifactId>
                        <version>8.14</version>
                    </dependency>
                </dependencies>
            </plugin>
```

**Pull request merged**